### PR TITLE
Geocoding: Set window size

### LINKED ISSUE #36: GeoCoding: change default to Encode
##### Geo version

0.2.3

##### Orange version

3.14.0

##### Expected behavior
GeoCoding encodes region names to coordinates on the fly.

##### Actual behavior
Default is decoding, which I think is not what most users want.

##### Suggestion
Change default to Encode. Search for string variables with names such as 'Country', 'State', 'Geo'... I am not sure about the heuristic. Should be similar to lat/lon detection. Alternatively, if no such variable is found, go with the first string variable. If no string variable found, go with the current default.

### PR DESCRIPTION
**Pull request by ajdapretnar**
##### Issue

Fixes #36 and window geometry.

##### Description of changes
Changed default to Encode.
Set window size depending on selection (hide when no unmatched instances).
Small cosmetic fixes.

##### Includes
- [X] Code changes
- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by ajdapretnar** [comment]
Don't yet merge this PR. Tests are pending + doing some other fixes, too.

**Comment by ajdapretnar** [comment]
This is ready for another review. I have set the widget to work with Encode by default and try to retrieve the first string or discrete variable in the data. If all continuous, switch to Decode. Tests added.

File: orangecontrib/geo/utils.py [hunk 1]
```diff
@@ -42,3 +43,14 @@ def _all_between(vals, min, max):
...
+def find_country(data):
+    """Return most likely region identifier"""
+    assert isinstance(data, Table)
+
+    str_attr = [meta for meta in data.domain.metas if isinstance(meta,
+                                                                 StringVariable)]
+    disc_attr = [disc for disc in data.domain.variables if isinstance(disc,
+                                                             DiscreteVariable)]
+
+    return list(chain(str_attr, disc_attr))
```
Comment (kernc) [inline orangecontrib/geo/utils.py]: `find_country`, yet the function returns a list of string metas and nominal variables?

It only ever needs the first element further down. Can avoid iterating over all variables with, better:
```py
def guess_region_attr_name(data: Table):
    """Return the name of the first variable that could specify a region name"""
    var = next(chain((var for var in data.domain.metas if var.is_string),
                     (var for var in data.domain.variables if var.is_discrete)),
               None)
    return var.name if var else None
```

File: orangecontrib/geo/widgets/owgeocoding.py [hunk 2]
```diff
@@ -272,19 +279,34 @@ def set_data(self, data):
...
             self.commit()
             return
 
         for model in self.domainmodels:
             model.set_domain(data.domain)
 
+        if find_country(data):
+            self.str_attr = find_country(data)[0].name
+        else:
+            self.str_attr = None
```
Comment (kernc) [inline orangecontrib/geo/widgets/owgeocoding.py]: And here just a single call:
```py
attr = self.str_attr = guess_region_attr_name(data)
if attr is None:
    self.is_decoding = 1
```

File: orangecontrib/geo/utils.py (hunk 1 above)
Comment (kernc) [reply orangecontrib/geo/utils.py]: Anyway, I'd put this very broad looking function inside owgecoding file. That, or inside Orange core. :)

Comment (ajdapretnar) [reply orangecontrib/geo/utils.py]: Yeah, I didn't know what to do with it. I'll move it to owgeocoding and implement suggestions.

Comment (astaric) [reply orangecontrib/geo/utils.py]: What about discrete vars in metas?

Comment (astaric) [reply orangecontrib/geo/utils.py]: naming the generators (storing them in variables) before being chained is also not a bad idea

I also sometimes find for loops with return easier to understand than next:

```
string_vars = (var for var in data.domain.metas if var.is_string)
discrete_vars = (var for var in data.domain.variables if var.is_discrete)
for var in chain(string_vars, discrete_vars):
    return var.name
```

as this is subjective, use whatever looks nice to you :)

**Comment by ajdapretnar** [comment]
Suggestions taken into consideration. Should work for string or discrete metas and discrete attributes.

File: orangecontrib/geo/widgets/owgeocoding.py [hunk 3]
```diff
@@ -26,12 +28,20 @@ def available_name(domain, template):
         if name not in domain:
             return name
 
+def guess_region_attr_name(data):
+    """Return the name of the first variable that could specify a region name"""
+    var = next(chain((var for var in data.domain.metas if var.is_string or
+                      var.is_discrete),
+                     (var for var in data.domain.variables if
+                      var.is_discrete)),
```
Comment (kernc) [inline orangecontrib/geo/widgets/owgeocoding.py]: It's customary to break long comprehension lines at (before) if.

For the record, @astaric's proposed version was better.

**Comment by ajdapretnar** [comment]
All done.

**Pull request merged**